README.md
# freqpcr
Interval estimation of population allele frequency based on quantitative PCR ΔΔCq measures from bulk samples

* [README in Japanese（日本語: equivalent to this document + vignette）](https://github.com/sudoms/freqpcr/blob/master/README.jp.md)

# Features
Interval estimation of the population allele frequency from qPCR analysis based on the restriction enzyme digestion (RED)-ΔΔCq method (Osakabe et al. 2017), as well as general  ΔΔCq analysis.
Compatible with the Cq measurement of DNA extracted from multiple individuals at once, so called ``group-testing'', this model assumes that the quantity of DNA extracted from an individual organism follows a gamma distribution.
Therefore, the point estimate is robust regarding the uncertainty of the DNA yield.

See the package vignette [Introduction to freqpcr] and [published article (Sudo and Osakabe 2021 Molecular Ecology Resources)](https://doi.org/10.1111/1755-0998.13554) for the model structure.

Prerequisites are:

* Want to know the percentage of the specific allele at the population level (population allele frequency) relative to other allele(s) on the same locus.

* For the population in question, multiple DNA solutions (bulk samples) have been obtained, each of which consists of multiple individuals extracted at once.

* Quantitative PCR based on the real-time PCR analysis has been performed on each bulk sample and a set of four Cq values is available.

* The ΔΔCq value calculated from the Cq quartet represents the allele frequency of the bulk sample (sample allele frequency).

The main function `freqpcr()` takes as input data the number of individuals included in each bulk sample (`N`) and the set of Cq values (`housek0`, `target0`, `housek1`, `target1`) measured for each bulk sample (since it is a ΔΔCq method, there are usually four Cq values).

Prior to the estimation with samples with unknown allele ratios, auxiliary experimental parameters e.g. amplification efficiency during real-time PCR are required. Functions for estimating experimental parameters (`knownqpcr()` and `knownqpcr_unpaired()` functions) are also provided, using multiple DNA solutions with known allele mixing ratios.

> Since v0.4.0, the `freqpcr()` function has been extended to accept datasets lacking control samples (`housek0` and `target0`) i.e., a ΔCq method provided the sizes of the auxiliary experimental parameters are known.


# Requirement
* R (>= 3.6)
* cubature (https://cran.r-project.org/package=cubature)

# Installation

```R
library(remotes)
install_github("sudoms/freqpcr")

library(freqpcr)
packageVersion("freqpcr")
```

If there are errors (converted from warning), it might be the case the dependent package 'cubature' has been built on a newer version of R (https://github.com/r-lib/remotes/issues/403).

```R
** byte-compile and prepare package for lazy loading
Error: (converted from warning) package 'cubature' was built under R version 3.6.3
```

Then, set the following environment variable: `R_REMOTES_NO_ERRORS_FROM_WARNINGS="true"` on you R session and run install_github() again.

```R
Sys.setenv(R_REMOTES_NO_ERRORS_FROM_WARNINGS="true")
install_github("sudoms/freqpcr")
```

# Usage

```R
library(freqpcr)
```

## First, define the parameters for dummy data generation

Dummy Cq data is generated for demo use. As for real qPCR experiments, you first need to estimate those auxiliary parameters using samples with known multiple allele-mixing ratios and `knownqpcr()` or `knownqpcr_unpaired()` functions. See the vignette for detail.

```R
P <- 0.15
K <- 4

# These auxiliary parameters must be pre-defined 
EPCR <- 0.97
zeroAmount <- 0.0016
scaleDNA <- 1e-06 # used in make_dummy() but not in freqpcr()
targetScale <- 1.2 
baseChange <- 0.2# used in make_dummy() but not in freqpcr()
sdMeasure <- 0.2
```

P:
* The population allele frequency to be estimated

K:
* The gamma shape parameter of the individual DNA yield

EPCR:
* Amplification efficiency per PCR cycle, given as a positive numeric.
* When EPCR = 1, template DNA doubles every cycle (EPCR + 1 = 2).

zeroAmount:
* (In RED-ΔΔCq method) residual rate of restriction enzyme digestion.
* (in general ΔΔCq and ΔCq analyses) small portion of the off-target allele amplified in the PCR process.
* It needs to be always specified by the user as a number between 0 and 1, usually near 0.

scaleDNA:
* Relative amount of the template DNA (of the housekeeping gene) compared to the threshold of realtime PCR. 
* If DNA increases twofold in a PCR cycle (`EPCR = 1`), `scaleDNA` = 10^-6 means Cq value is approximately 20.

targetScale:
* Relative DNA amount of the target locus compared with the housekeeping gene.

baseChange:
* Relative amount of the template DNA (of the housekeeping gene) before and after the restriction enzyme treatment in the RED-ΔΔCq method (as an involuntary change due to experimental manipulation).
* The variable does not exist in other common ΔΔCq and ΔCq methods.

sdMeasure:
* The measurement error on each Cq (Ct) value, following Normal(0, sdMeasure)

## Make a dummy Cq dataset

With four bulk samples, each of which comprises eight haploid individuals

```R
dmy_cq <- make_dummy(   rand.seed=71, P=P, K=K, ntrap=4, npertrap=8,
                        scaleDNA=scaleDNA, targetScale=targetScale, 
                        baseChange=targetScale, EPCR=EPCR, 
                        zeroAmount=zeroAmount,
                        sdMeasure=sdMeasure, diploid=FALSE  )
print(dmy_cq)
```

```R
> print(dmy_cq)
An object of class "CqList"
Slot "N":
[1] 8 8 8 8

Slot "m":
     [,1] [,2] [,3] [,4]
[1,]    1    1    1    0
[2,]    7    7    7    8

Slot "xR":
[1] 2.661931e-06 6.163087e-06 3.199384e-06 0.000000e+00

Slot "xS":
[1] 2.216434e-05 3.163697e-05 3.221190e-05 4.018694e-05

Slot "housek0":
[1] 15.52662 14.86406 15.09659 14.84089

Slot "target0":
[1] 15.61207 14.80267 14.53848 14.85648

Slot "housek1":
[1] 17.85924 17.04084 17.54948 17.38438

Slot "target1":
[1] 21.08799 19.55396 20.92745 26.29820

Slot "DCW":
[1]  0.08544953 -0.06139074 -0.55811003  0.01559466

Slot "DCD":
[1] 3.228745 2.513126 3.377968 8.913814

Slot "deldel":
[1] 3.143296 2.574517 3.936078 8.898219

Slot "RFreqMeasure":
[1] 0.11868765 0.17453873 0.06933585 0.00239759

Slot "ObsP":
[1] 0.11868765 0.17453873 0.06933585 0.00239759

Slot "rand.seed":
[1] 71
```

## Estimation with freqpcr

* When `P`, `K`, `sdMeasure`, and `targetScale` are marked unknown.

```R
result <- freqpcr(  N=dmy_cq@N, housek0=dmy_cq@housek0, target0=dmy_cq@target0,
                    housek1=dmy_cq@housek1, target1=dmy_cq@target1,
                    EPCR=EPCR, zeroAmount=zeroAmount, beta=TRUE, print.level=2  )
print(result)
```

```R
> print(result)
An object of class "CqFreq"
Slot "report":
                                                 Estimate Fixed  (scaled) (scaled.SE)       2.5%       97.5%
P (R-allele frequency)                         0.09773189     0 -2.222684  0.60914923 0.03178086   0.2633220
K (gamma shape parameter)                     20.92728471     0  3.041054  1.83522515 0.57354355 763.5884712
targetScale (relative amount of target locus)  1.11922896     0  0.112640  0.08911953 0.93985370   1.3328388
Cq measurement error (SD)                      0.20973065     0 -1.561931  0.32845070 0.11017528   0.3992451
EPCR (Duplication efficiency of PCR)           0.97000000     1        NA          NA         NA          NA

Slot "obj":
$minimum
[1] 6.094915

$estimate
[1] -2.222684  3.041054  0.112640 -1.561931

$gradient
[1] -3.400973e-05 -8.275889e-05 -5.170087e-05  8.878422e-05

$hessian
            [,1]        [,2]       [,3]       [,4]
[1,]  2.71023719  0.05094362   1.168535 -0.1766756
[2,]  0.05094362  0.37630056   2.045198 -0.6539723
[3,]  1.16853469  2.04519835 147.389558  6.4578190
[4,] -0.17667556 -0.65397228   6.457819 11.1504636

$code
[1] 1

$iterations
[1] 12


Slot "cal.time":
   user  system elapsed
   0.72    0.15    0.88
```
* When you have knowledge on the exact size of K (= set as a fixed parameter).

```R
result <- freqpcr(  N=dmy_cq@N, housek0=dmy_cq@housek0, target0=dmy_cq@target0,
                    housek1=dmy_cq@housek1, target1=dmy_cq@target1,
                    K=4,
                    EPCR=EPCR, zeroAmount=zeroAmount, beta=TRUE, print.level=1  )
```


# External links

* GitHub repository https://github.com/sudoms/freqpcr

* PDF help of the latest version https://github.com/sudoms/freqpcr/releases/latest

* Paper https://doi.org/10.1111/1755-0998.13554 
<br>
Sudo, M., & Osakabe, M. (2021). freqpcr: Estimation of population allele frequency using qPCR ΔΔCq measures from bulk samples. Molecular Ecology Resources, 00, 1–14. Creative Commons Attribution 4.0 International (CC BY 4.0)

* Author website https://sudori.info/english.html


# citation

```R
citation("freqpcr")
```


# Author

Masaaki Sudo (https://orcid.org/0000-0001-9834-9857)
National Agriculture and Food Research Organization (NARO), Japan


# License

GNU GPL (>= 3)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

# History

* NEWS.md


NEWS.md

# freqpcr 0.4.0

* First submitted version to CRAN

* The paper (Sudo and Osakabe 2021) was published online (https://doi.org/10.1111/1755-0998.13554 )

* Vignette has been added

## New features

* The `freqpcr()` function has been extended to accept datasets lacking control samples (`housek0` and `target0`)

# freqpcr 0.3.5 

## Bug fixes
* (2021.02.17) `Lapack routine dgesv: system is exactly singular` in `knownqpcr()` when Cq contained NAs

# freqpcr 0.3.4 

* (2021.02.16) Released version of 0.3.3 (the code did not change)

# freqpcr 0.3.3

## New features

* (2021.02.13) Continuous distribution in sample allele ratio was supported. `knownqpcr()` function now officially accepts missing Cq observations (NA)

# freqpcr 0.3.2 

* (2021.01.29) Fixed: R version 2.14 -> 3.6 and higher

# freqpcr 0.3.1

* (2021.01.20) First published (first draft of the preprint)


R/00_package.R
################################################################################################################################

# Name: freqpcr
# Short description: Interval estimation of population allele frequency based on quantitative PCR double-Delta Cq measures from bulk samples
# Author: Masaaki Sudo (NARO, JAPAN)
# Maintainer: Masaaki Sudo
## https://sudori.info/english.html
## https://orcid.org/0000-0001-9834-9857

# License:
# Copyright 2020 Masaaki Sudo
# The source code is distributed under the GNU General Public License version 3 (GPLv3)

## This program is free software: you can redistribute it and/or modify it under the terms of
## the GNU General Public License as published by the Free Software Foundation, either version 3 of the License,
## or (at your option) any later version.
## This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
## without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
## See the GNU General Public License for more details.
## You should have received a copy of the GNU General Public License along with this program.
## If not, see <http://www.gnu.org/licenses/>.

#' The freqpcr package
#'
#' Maximum Interval estimation of population allele frequency based on quantitative PCR double-Delta Cq measures from bulk samples.
#'
#' @docType package
#' @importFrom methods new
#' @importFrom cubature cubintegrate
#' @importFrom stats dbeta dbinom dgamma dmultinom dnorm nlm optim plogis qlogis qnorm rbinom rgamma rmultinom rnorm
#' @importFrom utils flush.console
#' @keywords internal
#' @encoding UTF-8
#' @aliases freqpcr-package
"_PACKAGE"


R/01_integrator.R
# 01 integrator

#' @title Integration of likelihood function based on the beta assumption.

#' @description Internal function to integrate the likelihood getting the \eqn{\Delta}Cq value (the argument \code{del}) over the entire range of the allele ratio (0 to 1). Vectorized for multiple bulk samples. Having the same arguments with .integrate_gamma.
#' @param del Numeric vector of the observed \eqn{\Delta}Cq values.
#' @param SHR The gamma shape parameters for the mutant (R) portion of the bulk samples. Should be the same vector length as del. Each element of SHR is defined as K*(the assumed number of R allele in the bulk sample: 1, 2, 3, ..., n-1).
#' @param SHS The gamma shape parameters for the wild (S) portion of the bulk samples. Should be the same length as del. Each element of SHS is defined as K*(the assumed number of S allele in the bulk sample).
#' @param xsm Specify the accumulation of the standard deviation of the Cq measuring errors when the *-Cq values are fed as difference. For \eqn{\Delta}Cq values, sdMeasure times two. For \eqn{\Delta\Delta}Cq, sdMeasure times four. Default is two and used in most cases.
#' @param cubmethod Cubature method passed to the integrator function. See the section "Methods for cubintegrate".
#' @param relTol The maximum tolerance passed to the cubature method. Though the default of cubature::cubintegrate function is 1e-5, the accuracy is reduced here to acceralate the integration.
#' @param absTol The absolute tolerance passed to the cubature method. The default is 1e-8, which is less accurate than the default of cubintegrate function (1e-12) but considered enough for the estimation.
#' @param maxEval Maximum number of function evaluations needed. The default is 10^6, which is same as the cubintegrate default.
#' @inheritParams freqpcr
#' @return A numeric vector of marginal likelihoods having the same length as \code{del}.
#' @family integrators
#' @section Methods for cubintegrate: The following methods are available for \code{\link[cubature]{cubintegrate}()}: cubmethod = c("hcubature", "pcubature", "cuhre", "divonne", "suave", "vegas").
#' \cr
#' \code{hcubature} is considerably fast, but less accurate with larger \code{reltol}. \code{cuhre} is moderately fast and the most accurate in most \code{relTol} range. If you can wait longer, \code{hcubature} with \code{relTol = 1e-4} or \code{cuhre} with \code{relTol = 1e-1} is recommended.
#' At \code{reltol = 1e-1}, \code{hcubature} is three times faster than \code{cuhre}, but the log-likelihood fluctuate by 0.1 (by 0.001 in \code{cuhre}).
#' The speed and accuracy of \code{hcubature} at \code{reltol = 1e-5} is comparable with \code{cuhre} at \code{reltol = 1e-1}.
#' \cr
#' \code{pcubature} and \code{divonne} frequently returns \code{NaN} and are not recommended.
#' \cr
#' \code{suave} and \code{vegas} are Monte Carlo integration and slow. They are certainly accurate even at \code{reltol = 1}, but interior to \code{cuhre} with same \code{reltol}.
#' @keywords internal
.integrate_beta <- function (   del, SHR, SHS, zeroAmount, targetScale, sdMeasure, xsm=2, EPCR,
                                cubmethod="hcubature", relTol=1e-1, absTol=1e-8, maxEval=10^6) {
    fa <- function(x, SHR, SHS, del, zeroAmount, targetScale, sdMeasure, xsm, EPCR) {
        dbeta(x, shape1=SHR, shape2=SHS) *
        dnorm(  del, 
                mean=-(log(targetScale)+log((zeroAmount+(1-zeroAmount)*x)))/log(1.0+EPCR), 
                sd=sqrt(xsm)*sdMeasure  )
    }
    return( cubintegrate(  f=fa, lower=0, upper=1, fDim=length(del),
                        method=cubmethod, relTol=relTol, absTol=absTol, maxEval=maxEval, nVec=1L,
                        SHR=SHR, SHS=SHS, del=del, zeroAmount=zeroAmount, targetScale=targetScale,
                        sdMeasure=sdMeasure, xsm=xsm, EPCR=EPCR)$integral )
}


#' @title Double integration of likelihood over the two DNA quantities obeying the gamma distributions.

#' @description Internal function to integrate the likelihood getting \eqn{\Delta}Cq value (the argument \code{del}) over the entire range of the DNA quantities of the two alleles, 0 <= x_S < Inf and 0 <= x_R < Inf. Vectorized for multiple bulk samples. It shares the arguments with .integrate_beta.
#' @inheritParams .integrate_beta
#' @return A numeric vector of marginal likelihoods having the same length as \code{del}.
#' @family integrators
#' @keywords internal
.integrate_gamma <- function (  del, SHR, SHS, zeroAmount, targetScale, sdMeasure, xsm=2, EPCR,
                                cubmethod="hcubature", relTol=1e-1, absTol=1e-8, maxEval=10^6) {
    # double integration of R and S DNA quantities  (x_R, x_S) over {0 to Inf, 0 to Inf}
    # transformed in a polar coordinates system (r, theta) over {0 to 1, 0 to pi/2}
    # xsm: For delta Cq, sdMeasure times two. For delta-delta Cq, sdMeasure times four.
    fa <- function(x=c(0, 0), SHR, SHS, del, zeroAmount, targetScale, sdMeasure, xsm, EPCR) {
        xS <- (x[1]/(1-x[1]))*cos(x[2]) # x[1] := r, x[2] := theta
        xR <- (x[1]/(1-x[1]))*sin(x[2])
        return(
            dgamma(xR, shape=SHR, scale=1) *
            dgamma(xS, shape=SHS, scale=1) *
            dnorm(  del,
                    mean=-(log(targetScale)+log(zeroAmount*xS+xR)-log(xS+xR))/log(1.0+EPCR),
                    sd=sqrt(xsm)*sdMeasure  ) *
            (x[1]/((1-x[1])^3))
        ) # when m=0 or m=n, SHR or SHS are always 0 and dgamma returns 0.
    }
    return( cubintegrate( f=fa, lower=c(0, 0), upper=c(1, pi/2), fDim=length(del),
                        method=cubmethod, relTol=relTol, absTol=absTol, maxEval=maxEval, nVec=1L,
                        SHR=SHR, SHS=SHS, del=del, zeroAmount=zeroAmount, targetScale=targetScale,
                        sdMeasure=sdMeasure, xsm=xsm, EPCR=EPCR)$integral )
}


R/04_simulation.R
# 04 simulation

#' @title S4 class storing the dummy Cq data for performance test.
#'
#' @description A dummy Cq dataset suitable for a package test, typically obtained as the output of \code{\link{make_dummy}()}.
#' @slot N Sample sizes as a numeric vector. The \code{ntrap} and  \code{npertrap} arguments of \code{\link{make_dummy}()} are inherited to the length of N and each \code{N[i]} element, the number of individuals (both for haploidy and diploidy) contained in the \emph{i}th bulk sample, respectively.
#' @slot m Segregation ratio. As for haploidy, \code{m} is a matrix with 2 rows and \code{ntrap} columns. \code{m[1, i]} and \code{m[2, i]} stores the number of R (mutant) or S (wild type) individuals while \code{N[i] = sum(m[, i])} specifies the total in the bulk sample. It has 3 rows and \code{ntrap} columns as for diploidy. While \code{m[1, i]} stands for the number of RR hogozygote individuals, \code{m[2, i]} and \code{m[3, i]} stand for the numbers of RS heterozygotes and SS homozygotes, respectively.
#' @slot xR,xS Numeric vector of the same length with N. \code{xR[i]} stores the amount of the template DNA for R allele contained in the \emph{i}th bulk sample.
#' @slot housek0,target0,housek1,target1 Numeric vectors of the same lengths with N. Store the generated Cq values.
#' @slot DCW \eqn{\Delta}Cq values measured on the control samples (DNA extract without endonuclease digestion in the RED-\eqn{\Delta\Delta}Cq method, or pure R solution in a general \eqn{\Delta\Delta}Cq method), \code{DCW}, is defined as (\code{target0 - housek0}).
#' @slot DCD \eqn{\Delta}Cq values measured on the test samples (samples after endonuclease digestion in the RED-\eqn{\Delta\Delta}Cq method, or samples with unknown allele mixing ratios in a general \eqn{\Delta\Delta}Cq method), \code{DCD}, is defined as (\code{target1 - housek1}).
#' @slot deldel \eqn{\Delta\Delta}Cq value, defined as (\code{DCD - DCW}).
#' @slot RFreqMeasure A classical index of the allele frequency calculated for each bulk sample, which is defined as \code{(1.0+EPCR)^(-deldel)}. Note that the values of \code{EPCR} and other parameters, such as \code{P} or \code{K}, are not recorded in the object to avoid leakage of information.
#' @slot ObsP As \code{RFreqMeasure} can exceed 1 by definition, \code{ObsP} is defined as \code{min(RFreqMeasure, 1)}.
#' @slot rand.seed The seed of the random-number generator (RNG) which was fed to the current R session to generate dummy \code{m}, \code{xR} and \code{xS} data.
#' @export
CqList <- setClass("CqList",
    slots = c(
        N="numeric",
        m="matrix",
        xR="numeric", xS="numeric",
        housek0="numeric", target0="numeric", housek1="numeric", target1="numeric",
        DCW="numeric", DCD="numeric",
        deldel="numeric",
        RFreqMeasure="numeric",
        ObsP="numeric",
        rand.seed="numeric"
    )
)


#' @title Generate dummy DNA dataset ready for allele-frequency estimation.
#'
#' @description The function generates a dummy dataset of typical RED-\eqn{\Delta\Delta}Cq analysis. You can directly feed the output of this function to the first argument of \code{\link{sim_dummy}()}.
#' @param rand.seed Seed for the R built-in random-number-generator.
#' @param P A numeric between 0 and 1 giving the population allele frequency from which the test samples are generated.
#' @param K A positive numeric of the gamma shape parameter of the individual DNA yield.
#' @param ntrap,npertrap Scalar specifying the number of bulk samples (\code{ntrap}) and the numbers of individuals contained in each bulk sample (\code{npertrap}). Currently limited to the cases that all bulk samples have the same sample size: e.g. (4 + 4 + 4) when \code{ntrap = 3} and \code{npertrap = 4} hold.
#' @param scaleDNA Small positive scalar that specifies the scale parameter of the gamma distribution appriximating the DNA yield from (per-haploid) individual. The yield of \code{2*scaleDNA} is expected from a diploid. The quantity is determined as the relative amount, in linear scale, to the termination threshold of the real-time PCR.
#' @param targetScale (\eqn{\delta_{T}}) The relative template DNA amount of the target gene to the houskeeping gene, given as a positive numeric.
#' @param baseChange (\eqn{\delta_{B}}) The change rate in the template DNA quantities after the restriction enzyme digestion (in the RED-\eqn{\Delta\Delta}Cq method), given as a positive numeric. This parameter is not used in \code{\link{freqpcr}()}.
#' @param EPCR (\eqn{\eta}) Amplification efficiency per PCR cycle, given as a positive numeric. When \code{EPCR = 1}, template DNA doubles every cycle (\code{EPCR + 1 = 2}).
#' @param zeroAmount A numeric between 0 and 1, usually near 0, giving the residue rate of restriction enzyme digestion in RED-\eqn{\Delta\Delta}Cq method.
#' @inheritParams freqpcr
#' @return Object of the S4 class \linkS4class{CqList}, storing the dummy experiment data of Cq-based qPCR analysis. Note that a \linkS4class{CqList} object in no way contains original information on \code{P}, \code{K}, \code{targetScale}, \code{sdMeasure}, and \code{EPCR}.
#' @examples
#' P <- 0.25
#' # Just a test: segregation ratios for six bulk samples, 1000 individuals for each.
#' rmultinom(n=6, size=1000, prob=c(P, 1-P)) # haploidy
#' rmultinom(6, size=1000, prob=c(P^2, 2*P*(1-P), (1-P)^2)) # diploidy
#'
#' # Dummy Cq dataset with six bulk samples, each of which comprises eight haploids.
#' dmy_cq <- make_dummy( rand.seed=1, P=0.15, K=2, ntrap=6, npertrap=8,
#'                       scaleDNA=1e-07, targetScale=1.5, baseChange=0.3, EPCR=0.95,
#'                       zeroAmount=1e-3, sdMeasure=0.3, diploid=FALSE )
#' print(dmy_cq)
#' @export
make_dummy <- function( rand.seed, P, K, ntrap, npertrap, scaleDNA=(1/K)*1e-06,
                        targetScale, baseChange, EPCR, zeroAmount, sdMeasure, diploid=FALSE ) {
    set.seed(rand.seed)
    N <- rep(npertrap, ntrap)
    if (!diploid) {
        m <- rmultinom(n=ntrap, size=npertrap, prob=c(P, 1-P)) # R and S in each trap
        xR <- rgamma(n=ntrap, shape=K*m[1, ], scale=scaleDNA) # first row of m: R
        xS <- rgamma(n=ntrap, shape=K*m[2, ], scale=scaleDNA) # second row of m: S
    } else {
        m <- rmultinom(n=ntrap, size=npertrap, prob=c(P^2, 2*P*(1-P), (1-P)^2)) # RR, RS, and SS in each trap
        mx <- matrix(rgamma(n=length(c(m)), shape=c(m)*K, scale=scaleDNA), byrow=FALSE, nrow=nrow(m))
        xS <- 2*mx[3, ]+mx[2, ]
        xR <- 2*mx[1, ]+mx[2, ]
    }
    housek0 <- rnorm(n=ntrap, mean=-log(xR+xS)/log(1.0+EPCR), sd=sdMeasure)
    target0 <- rnorm(n=ntrap, mean=-(log(targetScale) + log(xR+xS))/log(1.0+EPCR), sd=sdMeasure)
    housek1 <- rnorm(n=ntrap, mean=-(log(baseChange) + log(xR+xS))/log(1.0+EPCR), sd=sdMeasure)
    target1 <- rnorm(n=ntrap, mean=-(log(targetScale) + log(baseChange) + log(xR+zeroAmount*xS))/log(1.0+EPCR),
                        sd=sdMeasure)
    set.seed(NULL)
    DCW <- target0-housek0
    DCD <- target1-housek1
    deldel <- DCD - DCW
    RFreqMeasure <- (1.0+EPCR)^(-deldel)
    ObsP <- ifelse(RFreqMeasure>1, 1, RFreqMeasure)
    return( CqList( N=N, m=m,
                    xR=xR, xS=xS, housek0=housek0, target0=target0, housek1=housek1, target1=target1,
                    DCW=DCW, DCD=DCD, deldel=deldel,
                    RFreqMeasure=RFreqMeasure, ObsP=ObsP, rand.seed=rand.seed ) )
}



#' @title Simulate freqpcr estimation based on user-generated dummy data.
#'
#' @description Wrapper of \code{\link{freqpcr}()} suitable for the performance test using a randomly-generated data object.
#' @param CqList Object belonging to the \linkS4class{CqList} class, typically the output from \code{\link{make_dummy}()}. Having the slots \code{N}, \code{housek0}, \code{target0}, \code{housek1}, and \code{target1}, all of which are numeric vectors of the same length.
#' @param P,K,targetScale,sdMeasure If NULL (default), the parameter is considered unknown and estimated via \code{\link{freqpcr}()}. If a value is specified, it is passed to \code{\link{freqpcr}()} as a fixed parameter. On the contrary, \code{EPCR} and \code{zeroAmount} are always treated as fixed parameters, for which values must be supplied.
#' @param beta,diploid,maxtime,print.level Configuration parameters which are passed directly to \code{\link{freqpcr}()}.
#' @param aux Additional information to be displayed. The default is \code{NULL}. If some value is input by the user, it is echoed to stdout together with the contents of the argument \code{CqList}. This option is convenient when you want to record the original dummy dataset and the corresponding result sequentially e.g. using \code{capture.output()}.
#' @param ... Additional arguments passed to \code{\link{freqpcr}()}.
#' @inheritParams make_dummy
#' @return Object of the S4 class \linkS4class{CqFreq}, which is same as \code{\link{freqpcr}()}.
#' @examples
#' # Prepare the parameter values.
#' K <- 2 # You already know the size of K in this case.
#' EPCR <- 0.97 # The sizes of EPCR and zeroAmount must always be supplied.
#' zeroAmount <- 1.6e-03
#' is.diploid <- FALSE
#'
#' # First, make a dummy Cq dataset with six bulk DNA samples,
#' # each of which comprises of eight haploid individuals.
#' dmy_cq <- make_dummy( rand.seed=1, P=0.75, K=K, ntrap=6, npertrap=8, scaleDNA=1e-07,
#'                       targetScale=1.5, baseChange=0.3, EPCR=EPCR,
#'                       zeroAmount=zeroAmount, sdMeasure=0.3, diploid=is.diploid )
#'
#' # Estimate the population allele frequency on the dummy dataset, presupposing K = 2.
#' sim_dummy(  CqList=dmy_cq, EPCR=EPCR, zeroAmount=zeroAmount,
#'             K=K,
#'             beta=TRUE, diploid=is.diploid, maxtime=60, print.level=2, aux="test" )
#'
#' # If the maximum calculation time was too short to converge, nlm() returns error.
#' # Then sim_dummy() returns a matrix filled with zeros.
#' sim_dummy(  CqList=dmy_cq, EPCR=EPCR, zeroAmount=zeroAmount,
#'             beta=TRUE, diploid=is.diploid, maxtime=0.01, print.level=2 )
#' @export
#' @family estimation procedures
sim_dummy <- function(  CqList, EPCR, zeroAmount,
                        P=NULL, K=NULL, targetScale=NULL, sdMeasure=NULL,
                        beta, diploid, maxtime, print.level, aux=NULL, ...  ) {
    if (!is.null(aux)) {
        cat("\n\n")
        cat("--------------------------------------------------------------------------------")
        cat("\n")
        print(c(PID=Sys.getpid()), quote=FALSE)
        cat("\n")
        cat(aux)
        cat("\n\n")
        print(CqList)
        cat("\n")
        print(Sys.time(), quote=FALSE)
    }
    e1 <- try( {
        result.list <- freqpcr( N=CqList@N, # number of individuals (not chromosome sets)
                                housek0=CqList@housek0, target0=CqList@target0,
                                housek1=CqList@housek1, target1=CqList@target1,
                                P=P, K=K, targetScale=targetScale, sdMeasure=sdMeasure, EPCR=EPCR,
                                zeroAmount=zeroAmount, beta=beta, diploid=diploid,
                                # pvalue=0.05, gradtol=1e-4, steptol=1e-7, iterlim=100,
                                maxtime=maxtime,
                                print.level=print.level, ... )
    } )
    if (class(e1)=="try-error") {
        pvalue <- 0.05
        result <- matrix(numeric(5*6), byrow=FALSE, ncol=6) # nrow == length(param0.full)
        rownames(result) <- c(  "P (R-allele frequency)",
                                "K (gamma shape parameter)",
                                "targetScale (rel. DNA quantity)",
                                "sdMeasure (Cq measurement error)",
                                "EPCR (amplification per cycle)")
        colnames(result) <- c(  "Estimate", "Fixed", "(scaled)", "(scaled.SE)",
                                paste(deparse(100*   pvalue/2) , "%", sep=""),
                                paste(deparse(100*(1-pvalue/2)), "%", sep="") )
        ptime0 <- proc.time()
        cal.time <- proc.time()-ptime0
        result.list <- CqFreq(report=result, obj=list(iterations=NA), cal.time=cal.time)
    }
    print(Sys.time(), quote=FALSE)
    rm(e1)
    gc(); gc();
    return(result.list)
}
